src/JsonSchema/RefResolver.php

```diff
@@ -101,47 +117,66 @@ public function getUriRetriever()
      */
     public function resolve($schema, $sourceUri = null)
     {
-        if (self::$depth > self::$maxDepth) {
-            self::$depth = 0;
-            throw new JsonDecodingException(JSON_ERROR_DEPTH);
-        }
-        ++self::$depth;
-
-        if (! is_object($schema)) {
-            --self::$depth;
+        if (!is_object($schema)) {
             return;
         }
 
-        if (null === $sourceUri && ! empty($schema->id)) {
-            $sourceUri = $schema->id;
+        // Fill in id property
+        if ($sourceUri) {
+            $schema->id = $sourceUri;
         }
 
-        if (null === $this->rootSchema) {
-            $this->rootSchema = $schema;
-        }
-
-        // Resolve $ref first
-        $this->resolveRef($schema, $sourceUri);
+        // First determine our resolution scope
+        $scope = $this->getResolutionScope($schema, $sourceUri);
 
         // These properties are just schemas
         // eg.  items can be a schema or an array of schemas
         foreach (array('additionalItems', 'additionalProperties', 'extends', 'items') as $propertyName) {
-            $this->resolveProperty($schema, $propertyName, $sourceUri);
+            $this->resolveProperty($schema, $propertyName, $scope);
         }
 
         // These are all potentially arrays that contain schema objects
         // eg.  type can be a value or an array of values/schemas
         // eg.  items can be a schema or an array of schemas
         foreach (array('disallow', 'extends', 'items', 'type', 'allOf', 'anyOf', 'oneOf') as $propertyName) {
-            $this->resolveArrayOfSchemas($schema, $propertyName, $sourceUri);
+            $this->resolveArrayOfSchemas($schema, $propertyName, $scope);
         }
 
         // These are all objects containing properties whose values are schemas
-        foreach (array('dependencies', 'patternProperties', 'properties') as $propertyName) {
-            $this->resolveObjectOfSchemas($schema, $propertyName, $sourceUri);
+        foreach (array('definitions', 'dependencies', 'patternProperties', 'properties') as $propertyName) {
+            $this->resolveObjectOfSchemas($schema, $propertyName, $scope);
         }
 
-        --self::$depth;
+        // Resolve $ref
+        $this->resolveRef($schema, $scope);
```

Comment:
The reason for doing it afterward is to ensure we don't get stuck in an infinite loop whilst attempting to dereference everything.